File: Python/ProjetoIptvManterLinkUsuarioAtualizado/pythonProject/_Biblioteca.py
import requests
import base64
import time
import pandas as pd

# Declaração das variáveis globais
link_atualizado_tvs = ''
link_atualizado_uniplay = ''
link_atualizado_bit = ''
link_atualizado_fast = ''

sheet_url_links = "https://docs.google.com/spreadsheets/d/1lMIq91MwJcxuNDZsJaTLdYmj_M2bmLO5_R9Y5EWDEb8/export?format=csv"

# Configuração da API Key do 2Captcha
captcha_api_key = ''
# Função para resolver o CAPTCHA usando 2Captcha
def solve_captcha_with_2captcha(image_path, api_key, captcha_type='base64', **kwargs):
    """
    Resolve CAPTCHAs usando o serviço 2Captcha.

    :param image_path: Caminho para a imagem do CAPTCHA (para base64).
    :param api_key: Chave da API do 2Captcha.
    :param captcha_type: Tipo do CAPTCHA (base64, userrecaptcha, hcaptcha, etc.).
    :param kwargs: Parâmetros adicionais necessários para tipos específicos de CAPTCHA.
    :return: A resposta do CAPTCHA resolvido.
    """
    data = {
        'key': api_key,
        'method': captcha_type,
        'json': 1
    }

    # Para 'base64', envie a imagem em base64
    if captcha_type == 'base64':
        with open(image_path, "rb") as img_file:
            base64_image = base64.b64encode(img_file.read()).decode('utf-8')
            data['body'] = base64_image

    # Adicione parâmetros adicionais (ex.: googlekey, sitekey, etc.)
    data.update(kwargs)

    # Enviar solicitação para 2Captcha
    response = requests.post("http://2captcha.com/in.php", data=data)
    request_result = response.json()

    if request_result.get("status") != 1:
        raise Exception(f"Erro ao enviar CAPTCHA para 2Captcha: {request_result.get('request')}")

    # Pegar o ID do CAPTCHA
    captcha_id = request_result.get("request")

    # Consultar o resultado até ser resolvido
    result_url = f"http://2captcha.com/res.php?key={api_key}&action=get&id={captcha_id}&json=1"
    while True:
        time.sleep(5)  # Esperar 5 segundos entre tentativas
        result_response = requests.get(result_url)
        result_json = result_response.json()

        if result_json.get("status") == 1:
            return result_json.get("request")  # Resultado do CAPTCHA

        if result_json.get("request") != "CAPCHA_NOT_READY":
            raise Exception(f"Erro ao resolver CAPTCHA: {result_json.get('request')}")

# Lê a planilha usando pandas
# Lê a planilha usando pandas
def getGoogleSheetData(url):
    df = pd.read_csv(url, sep=',', quotechar='"', dtype=str).dropna(how="all")  # Remove linhas totalmente vazias
    return df


# Lê a planilha e atualiza os links de acordo com o "Nome painel"
def atualizarLinks():
    global link_atualizado_tvs, link_atualizado_uniplay, link_atualizado_bit, link_atualizado_fast  # Declara que as variáveis são globais


    df = getGoogleSheetData(sheet_url_links)

    # Dicionário para armazenar os links atualizados
    links_atualizados = {}

    # Itera sobre cada linha da planilha
    for index, row in df.iterrows():
        nome_painel = row['Nome painel']
        link = row['Link']

        # Verifica qual variável deve ser atualizada com base no nome do painel
        if nome_painel == 'link_tvs':
            links_atualizados['tvs'] = link
        elif nome_painel == 'link_uniplay':
            links_atualizados['uniplay'] = link
        elif nome_painel == 'link_bit':
            links_atualizados['bit'] = link
        elif nome_painel == 'link_fast':
            links_atualizados['fast'] = link

    # Atualiza as variáveis globais com os links extraídos da planilha e remove espaços
    link_atualizado_tvs = links_atualizados.get('tvs', '').strip()
    link_atualizado_uniplay = links_atualizados.get('uniplay', '').strip()
    link_atualizado_bit = links_atualizados.get('bit', '').strip()
    link_atualizado_fast = links_atualizados.get('fast', '').strip()

    print("Links atualizados com sucesso:")
    print(f"TVS: {link_atualizado_tvs}")
    print(f"Uniplay: {link_atualizado_uniplay}")
    print(f"Bit: {link_atualizado_bit}")
    print(f"Fast: {link_atualizado_fast}")

def obterLinkAtualizado(pServidor: str) -> str:
    if pServidor == 'TVS':
        return link_atualizado_tvs
    elif pServidor == 'UNIPLAY':
        return link_atualizado_uniplay
    elif pServidor == 'BIT':
        return link_atualizado_bit
    elif pServidor == 'FAST':
        return link_atualizado_fast
    else:
        return ""

File: JavaScript/ProjetoIptvManterLinkPlaywright/src/biblioteca.js
const axios = require('axios');
const { parse } = require('csv-parse/sync');

// Variáveis globais dos links
let linkAtualizadoTvs = '';
let linkAtualizadoUniplay = '';
let linkAtualizadoBit = '';
let linkAtualizadoFast = '';

let captchaApiKey = '';

const sheetUrlLinks =
  'https://docs.google.com/spreadsheets/d/1lMIq91MwJcxuNDZsJaTLdYmj_M2bmLO5_R9Y5EWDEb8/export?format=csv';

async function getGoogleSheetData(url) {
  const response = await axios.get(url);
  const records = parse(response.data, {
    columns: true,
    skip_empty_lines: true,
    trim: true,
  });
  return records;
}

async function atualizarLinks() {
  const data = await getGoogleSheetData(sheetUrlLinks);

  for (const row of data) {
    const nomePainel = (row['Nome painel'] || '').trim();
    const link = (row['Link'] || '').trim();

    if (nomePainel === 'link_tvs') linkAtualizadoTvs = link;
    else if (nomePainel === 'link_uniplay') linkAtualizadoUniplay = link;
    else if (nomePainel === 'link_bit') linkAtualizadoBit = link;
    else if (nomePainel === 'link_fast') linkAtualizadoFast = link;
  }

  console.log('Links atualizados com sucesso:');
  console.log(`TVS: ${linkAtualizadoTvs}`);
  console.log(`Uniplay: ${linkAtualizadoUniplay}`);
  console.log(`Bit: ${linkAtualizadoBit}`);
  console.log(`Fast: ${linkAtualizadoFast}`);
}

function obterLinkAtualizado(servidor) {
  switch (servidor) {
    case 'TVS': return linkAtualizadoTvs;
    case 'UNIPLAY': return linkAtualizadoUniplay;
    case 'BIT': return linkAtualizadoBit;
    case 'FAST': return linkAtualizadoFast;
    default: return '';
  }
}

// Resolve CAPTCHA via 2Captcha enviando imagem em base64
async function solveCaptchaWith2Captcha(imageBuffer, apiKey) {
  const base64Image = imageBuffer.toString('base64');

  const params = new URLSearchParams();
  params.append('key', apiKey);
  params.append('method', 'base64');
  params.append('body', base64Image);
  params.append('json', '1');

  const response = await axios.post('http://2captcha.com/in.php', params);

  if (response.data.status !== 1) {
    throw new Error(`Erro ao enviar CAPTCHA para 2Captcha: ${response.data.request}`);
  }

  const captchaId = response.data.request;
  const resultUrl = `http://2captcha.com/res.php?key=${apiKey}&action=get&id=${captchaId}&json=1`;

  while (true) {
    await new Promise((resolve) => setTimeout(resolve, 5000));
    const result = await axios.get(resultUrl);

    if (result.data.status === 1) {
      return result.data.request;
    }

    if (result.data.request !== 'CAPCHA_NOT_READY') {
      throw new Error(`Erro ao resolver CAPTCHA: ${result.data.request}`);
    }
  }
}

module.exports = {
  getGoogleSheetData,
  atualizarLinks,
  obterLinkAtualizado,
  solveCaptchaWith2Captcha,
  get captchaApiKey() { return captchaApiKey; },
  set captchaApiKey(val) { captchaApiKey = val; },
};
